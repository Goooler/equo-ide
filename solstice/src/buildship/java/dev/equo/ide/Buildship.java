/*******************************************************************************
 * Copyright (c) 2022 EquoTech, Inc. and others.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     EquoTech, Inc. - initial API and implementation
 *******************************************************************************/
package dev.equo.ide;

import org.eclipse.buildship.core.GradleBuild;
import org.eclipse.buildship.core.GradleCore;
import org.eclipse.buildship.core.GradleDistribution;
import org.eclipse.buildship.ui.internal.util.gradle.GradleDistributionViewModel;
import org.eclipse.buildship.ui.internal.wizard.project.ProjectImportConfiguration;
import org.osgi.framework.BundleContext;

class Buildship implements IdeHookInstantiated {
	IdeHookBuildship data;

	/**
	 * Logic comes from {@link org.eclipse.buildship.ui.internal.wizard.project.ProjectImportWizard}
	 */
	Buildship(IdeHookBuildship data) {
		this.data = data;
	}

	boolean isClean;

	/**
	 * The very first method to be called, called as soon as the command line arguments have been
	 * parsed.
	 */
	@Override
	public void isClean(boolean isClean) {
		this.isClean = isClean;
	}

	BundleContext context;

	/** Called after the OSGi container is created and populated. */
	@Override
	public void afterOsgi(BundleContext context) {
		this.context = context;
	}

	/**
	 * This method is called during workbench initialization prior to any windows being opened.
	 *
	 * <p>{@link org.eclipse.ui.application.WorkbenchAdvisor#initialize}
	 */
	@Override
	public void initialize() {
		if (isClean) {
			// we only need to import the project when the IDE is fresh
			return;
		}
		var configuration = new ProjectImportConfiguration();
		configuration.setProjectDir(data.rootDir);
		configuration.setOverwriteWorkspaceSettings(false);
		configuration.setDistribution(GradleDistributionViewModel.from(GradleDistribution.fromBuild()));
		configuration.setOfflineMode(data.isOffline);

		var internalBuildConfiguration = configuration.toInternalBuildConfiguration();
		boolean showExecutionsView =
				internalBuildConfiguration.getWorkspaceConfiguration().isShowExecutionsView();
		var buildConfiguration = configuration.toBuildConfiguration();
		GradleBuild gradleBuild = GradleCore.getWorkspace().createBuild(buildConfiguration);

		// TODO: very incomplete.
	}
}
