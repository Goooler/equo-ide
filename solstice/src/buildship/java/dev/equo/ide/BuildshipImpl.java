/*******************************************************************************
 * Copyright (c) 2022 EquoTech, Inc. and others.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     EquoTech, Inc. - initial API and implementation
 *******************************************************************************/
package dev.equo.ide;

import com.google.common.collect.ImmutableList;
import java.lang.reflect.InvocationTargetException;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import org.eclipse.buildship.core.GradleCore;
import org.eclipse.buildship.core.GradleDistribution;
import org.eclipse.buildship.core.SynchronizationResult;
import org.eclipse.buildship.core.internal.DefaultGradleBuild;
import org.eclipse.buildship.core.internal.workspace.NewProjectHandler;
import org.eclipse.buildship.ui.internal.util.gradle.GradleDistributionViewModel;
import org.eclipse.buildship.ui.internal.util.workbench.WorkbenchUtils;
import org.eclipse.buildship.ui.internal.util.workbench.WorkingSetUtils;
import org.eclipse.buildship.ui.internal.wizard.project.ProjectImportConfiguration;
import org.eclipse.core.resources.IProject;
import org.eclipse.core.runtime.CoreException;
import org.eclipse.core.runtime.IProgressMonitor;
import org.eclipse.swt.widgets.Display;
import org.eclipse.ui.IWorkingSet;
import org.eclipse.ui.IWorkingSetManager;
import org.eclipse.ui.PlatformUI;
import org.eclipse.ui.internal.Workbench;
import org.gradle.tooling.GradleConnector;
import org.osgi.framework.BundleContext;

class BuildshipImpl implements IdeHookInstantiated {
	IdeHookBuildship data;

	/**
	 * Logic comes from {@link org.eclipse.buildship.ui.internal.wizard.project.ProjectImportWizard}
	 */
	BuildshipImpl(IdeHookBuildship data) {
		this.data = data;
	}

	boolean isClean;

	/**
	 * The very first method to be called, called as soon as the command line arguments have been
	 * parsed.
	 */
	@Override
	public void isClean(boolean isClean) {
		this.isClean = isClean;
	}

	BundleContext context;

	/** Called after the OSGi container is created and populated. */
	@Override
	public void afterOsgi(BundleContext context) {
		this.context = context;
	}

	/**
	 * This method is called during workbench initialization prior to any windows being opened.
	 *
	 * <p>{@link org.eclipse.ui.application.WorkbenchAdvisor#initialize}
	 */
	@Override
	public void postStartup() throws Exception {
		if (!isClean) {
			// we only need to import the project when the IDE is fresh
			return;
		}
		Workbench.getInstance().getActiveWorkbenchWindow().run(true, true, this::doImport);
	}

	private void doImport(IProgressMonitor monitor) throws InvocationTargetException {
		var importCfg = new ProjectImportConfiguration();
		importCfg.setProjectDir(data.rootDir);
		importCfg.setOfflineMode(data.isOffline);

		var distribution = GradleDistribution.fromBuild();
		importCfg.setDistribution(GradleDistributionViewModel.from(distribution));

		importCfg.setAutoSync(true);
		importCfg.setApplyWorkingSets(true);
		importCfg.setBuildScansEnabled(false);
		importCfg.setShowConsoleView(true);
		importCfg.setShowExecutionsView(true);
		importCfg.setOverwriteWorkspaceSettings(true);
		importCfg.setWorkingSets(Collections.emptyList());

		var buildCfg = importCfg.toBuildConfiguration();
		var gradleBuild = GradleCore.getWorkspace().createBuild(buildCfg);
		var workingSetsAddingNewProjectHandler =
				new ImportWizardNewProjectHandler(
						NewProjectHandler.IMPORT_AND_MERGE,
						importCfg,
						importCfg.getShowExecutionsView().getValue());
		SynchronizationResult result =
				((DefaultGradleBuild) gradleBuild)
						.synchronize(
								workingSetsAddingNewProjectHandler,
								GradleConnector.newCancellationTokenSource(),
								monitor);
		if (!result.getStatus().isOK()) {
			throw new InvocationTargetException(new CoreException(result.getStatus()));
		}
	}

	private static final class ImportWizardNewProjectHandler implements NewProjectHandler {
		private final ProjectImportConfiguration configuration;
		private final NewProjectHandler importedBuildDelegate;
		private final boolean showExecutionsView;
		private volatile boolean gradleViewsVisible;

		private ImportWizardNewProjectHandler(
				NewProjectHandler delegate,
				ProjectImportConfiguration configuration,
				boolean showExecutionsView) {
			this.importedBuildDelegate = delegate;
			this.configuration = configuration;
			this.showExecutionsView = showExecutionsView;
		}

		public boolean shouldImportNewProjects() {
			return this.importedBuildDelegate.shouldImportNewProjects();
		}

		public void afterProjectImported(IProject project) {
			this.importedBuildDelegate.afterProjectImported(project);
			this.addWorkingSets(project);
			this.ensureGradleViewsAreVisible();
		}

		private void addWorkingSets(IProject project) {
			List<String> workingSetNames =
					configuration.getApplyWorkingSets().getValue()
							? ImmutableList.copyOf(
									(Collection<String>) this.configuration.getWorkingSets().getValue())
							: ImmutableList.of();
			IWorkingSetManager workingSetManager = PlatformUI.getWorkbench().getWorkingSetManager();
			IWorkingSet[] workingSets = WorkingSetUtils.toWorkingSets(workingSetNames);
			workingSetManager.addToWorkingSets(project, workingSets);
		}

		private void ensureGradleViewsAreVisible() {
			Display.getDefault()
					.asyncExec(
							() -> {
								WorkbenchUtils.showView(
										"org.eclipse.buildship.ui.views.taskview", (String) null, 1);
								WorkbenchUtils.showView(
										"org.eclipse.buildship.ui.views.executionview", (String) null, 2);
							});
		}
	}
}
