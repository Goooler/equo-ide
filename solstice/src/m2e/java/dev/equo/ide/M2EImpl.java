/*******************************************************************************
 * Copyright (c) 2023 EquoTech, Inc. and others.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     EquoTech, Inc. - initial API and implementation
 *******************************************************************************/
package dev.equo.ide;

import java.io.File;
import java.lang.reflect.InvocationTargetException;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import org.eclipse.core.runtime.CoreException;
import org.eclipse.core.runtime.IProgressMonitor;
import org.eclipse.m2e.core.MavenPlugin;
import org.eclipse.m2e.core.embedder.MavenModelManager;
import org.eclipse.m2e.core.project.LocalProjectScanner;
import org.eclipse.m2e.core.project.MavenProjectInfo;
import org.eclipse.m2e.core.project.ProjectImportConfiguration;
import org.eclipse.m2e.core.ui.internal.wizards.ImportMavenProjectsJob;
import org.eclipse.ui.IWorkingSet;
import org.eclipse.ui.internal.Workbench;
import org.osgi.framework.BundleContext;

class M2EImpl implements IdeHookInstantiated {
	IdeHookM2E data;

	M2EImpl(IdeHookM2E data) {
		this.data = data;
	}

	boolean isClean;

	/**
	 * The very first method to be called, called as soon as the command line arguments have been
	 * parsed.
	 */
	@Override
	public void isClean(boolean isClean) {
		this.isClean = isClean;
	}

	BundleContext context;

	/** Called after the OSGi container is created and populated. */
	@Override
	public void afterOsgi(BundleContext context) {
		this.context = context;
	}

	/**
	 * This method is called during workbench initialization prior to any windows being opened.
	 *
	 * <p>{@link org.eclipse.ui.application.WorkbenchAdvisor#initialize}
	 */
	@Override
	public void postStartup() throws Exception {
		if (!isClean) {
			// we only need to import the project when the IDE is fresh
			return;
		}
		Workbench.getInstance().getActiveWorkbenchWindow().run(true, true, this::doImport);
	}

	private LocalProjectScanner instantiateScanner() {
		boolean basedirRemameRequired = false;
		var list = List.of(data.rootDir.getAbsolutePath());
		try {
			try {
				// 2.1.2 (Java 17+)
				var c =
						LocalProjectScanner.class.getConstructor(
								List.class, boolean.class, MavenModelManager.class);
				return c.newInstance(list, basedirRemameRequired, MavenPlugin.getMavenModelManager());
			} catch (NoSuchMethodException e) {
				// 1.20.1 (Java 11+)
				var c =
						LocalProjectScanner.class.getConstructor(
								File.class, List.class, boolean.class, MavenModelManager.class);
				return c.newInstance(
						data.rootDir, list, basedirRemameRequired, MavenPlugin.getMavenModelManager());
			}
		} catch (InvocationTargetException
				| InstantiationException
				| IllegalAccessException
				| NoSuchMethodException e) {
			throw new RuntimeException(e);
		}
	}

	private void doImport(IProgressMonitor monitor)
			throws InvocationTargetException, InterruptedException {
		var discoverMonitor = monitor.slice(20);
		var importMonitor = monitor.slice(80);

		var scanner = instantiateScanner();
		scanner.run(discoverMonitor);

		var importConfiguration = new ProjectImportConfiguration();
		var workingSets = List.<IWorkingSet>of();

		var projects = new HashSet<MavenProjectInfo>();
		for (var p : scanner.getProjects()) {
			addAll(p, projects);
		}
		var job = new ImportMavenProjectsJob(projects, workingSets, importConfiguration);
		var status = job.runInWorkspace(importMonitor);
		if (!status.isOK()) {
			throw new InvocationTargetException(new CoreException(status));
		}
		monitor.done();
	}

	private void addAll(MavenProjectInfo proj, Collection<MavenProjectInfo> projects) {
		projects.add(proj);
		for (var child : proj.getProjects()) {
			addAll(child, projects);
		}
	}
}
