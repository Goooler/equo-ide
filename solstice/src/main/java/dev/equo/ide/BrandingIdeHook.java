/*******************************************************************************
 * Copyright (c) 2022 EquoTech, Inc. and others.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     EquoTech, Inc. - initial API and implementation
 *******************************************************************************/
package dev.equo.ide;

import java.util.Arrays;
import org.eclipse.swt.SWT;
import org.eclipse.swt.widgets.Display;
import org.eclipse.swt.widgets.Shell;

public class BrandingIdeHook implements IdeHook {
	@Override
	public IdeHookInstantiated instantiate() {
		return new Instantiated();
	}

	class Instantiated implements IdeHookInstantiated {
		private Shell splash;

		@Override
		public void afterDisplay(org.eclipse.swt.widgets.Display display) {
			var cursor = display.getCursorLocation();
			var monitors = display.getMonitors();
			var bestMonitor =
					Arrays.stream(monitors)
							.filter(monitor -> monitor.getBounds().contains(cursor))
							.findAny()
							.orElse(monitors[0])
							.getBounds();

			int imgX = 1000;
			int imgY = 1000;
			splash = new Shell(display, SWT.NO_TRIM);
			splash.setText("Branding");
			splash.setBounds(
					bestMonitor.x + (bestMonitor.width - imgX) / 2,
					bestMonitor.y + (bestMonitor.height - imgY) / 2,
					imgX,
					imgY);
			splash.setBackground(display.getSystemColor(SWT.COLOR_GREEN));
			splash.open();

			Display.setAppName("Branding");
			while (display.readAndDispatch())
				// pump the event loop enough to show the branding
				;
		}

		@Override
		public void postStartup() {
			splash.dispose();
			splash = null;

			Shell[] shells = Display.getCurrent().getShells();
			for (var shell : shells) {
				shell.setText("Branding");
			}
		}
	}
}
