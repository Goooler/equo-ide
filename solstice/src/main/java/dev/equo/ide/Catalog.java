/*******************************************************************************
 * Copyright (c) 2022 EquoTech, Inc. and others.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     EquoTech, Inc. - initial API and implementation
 *******************************************************************************/
package dev.equo.ide;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import javax.annotation.Nullable;

public class Catalog implements Comparable<Catalog> {
	private static String V = "${VERSION}";
	public static final Catalog PLATFORM =
			new Catalog(
					"platform",
					"https://download.eclipse.org/eclipse/updates/" + V,
					"4.26",
					"org.eclipse.platform.ide.categoryIU");
	public static final Catalog JDT =
			new Catalog("jdt", PLATFORM, "org.eclipse.releng.java.languages.categoryIU", PLATFORM);
	public static final Catalog GRADLE_BUILDSHIP =
			new Catalog(
					"gradleBuildship",
					"https://download.eclipse.org/buildship/updates/e423/releases/3.x/" + V,
					"3.1.6.v20220511-1359",
					"org.eclipse.buildship.feature.group",
					JDT);

	private final String name;
	private final String p2urlTemplate;
	private final String latestVersion;
	private final String toInstall;
	private final List<Catalog> requires;

	Catalog(String name, Catalog copyFrom, String toInstall, Catalog... requires) {
		this(name, copyFrom.p2urlTemplate, copyFrom.latestVersion, toInstall, requires);
	}

	Catalog(
			String name,
			String p2urlTemplate,
			String latestVersion,
			String toInstall,
			Catalog... requires) {
		this.name = name;
		if (p2urlTemplate.endsWith("/")) {
			this.p2urlTemplate = p2urlTemplate;
		} else {
			this.p2urlTemplate = p2urlTemplate + "/";
		}
		this.latestVersion = latestVersion;
		this.toInstall = toInstall;
		this.requires = Arrays.asList(requires);
	}

	public String getP2UrlTemplate() {
		return p2urlTemplate;
	}

	public String getName() {
		return name;
	}

	public List<Catalog> getRequires() {
		return requires;
	}

	public String getUrlForOverride(@Nullable String override) {
		if (override == null) {
			return p2urlTemplate.replace(V, latestVersion);
		} else if (isUrl(override)) {
			return override;
		} else {
			return p2urlTemplate.replace(V, override);
		}
	}

	public List<String> getTargetsFor(@Nullable String override) {
		return Collections.singletonList(toInstall);
	}

	public static boolean isUrl(String maybeUrl) {
		return maybeUrl.startsWith("https:") || maybeUrl.startsWith("http:");
	}

	@Override
	public int compareTo(Catalog o) {
		return name.compareTo(o.name);
	}

	@Override
	public int hashCode() {
		return name.hashCode();
	}

	@Override
	public boolean equals(Object obj) {
		if (obj instanceof Catalog) {
			return ((Catalog) obj).name.equals(name);
		} else {
			return false;
		}
	}
}
