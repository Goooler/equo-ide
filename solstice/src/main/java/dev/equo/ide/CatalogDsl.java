/*******************************************************************************
 * Copyright (c) 2022 EquoTech, Inc. and others.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     EquoTech, Inc. - initial API and implementation
 *******************************************************************************/
package dev.equo.ide;

import dev.equo.solstice.p2.P2Model;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.TreeMap;
import javax.annotation.Nullable;

/**
 * Base class for DSL configuring an {@link Catalog} for inclusion in the Gradle or Maven plugins.
 */
public class CatalogDsl {
	protected final Catalog catalog;
	private final @Nullable CatalogDsl addedAsTransitiveOf;
	private @Nullable String urlOverride;
	private final WorkspaceInit workspaceInit = new WorkspaceInit();

	protected CatalogDsl(Catalog catalog) {
		this(catalog, null);
	}

	private CatalogDsl(Catalog catalog, @Nullable CatalogDsl addedAsTransitiveOf) {
		this.catalog = catalog;
		this.addedAsTransitiveOf = addedAsTransitiveOf;
	}

	protected WorkspaceInit workspaceInit() {
		return workspaceInit;
	}

	/**
	 * Subclasses can override this method to include their IdeHooks in {@link
	 * TransitiveAwareList#putInto(P2Model, dev.equo.ide.IdeHook.List)}.
	 */
	protected List<IdeHook> ideHooks() {
		return List.of();
	}

	private String url() {
		return catalog.getUrlForOverride(urlOverride);
	}

	private List<String> installs() {
		return catalog.getTargetsFor(urlOverride);
	}

	private Map<String, P2Model.Filter> filters() {
		return catalog.getFiltersFor(urlOverride);
	}

	/** Sets the URL and/or version override. */
	protected void setUrlOverride(String urlOverride) {
		this.urlOverride = urlOverride;
	}

	/**
	 * If this catalog has the same URL template as one of its transitive dependencies, then it should
	 * also have the exact same urlOverride. If it doesn't, and one of them is null, then we can
	 * automatically override the null with the non-null to resolve the issue.
	 */
	void syncUrlWith(CatalogDsl other) {
		if (!catalog.getP2UrlTemplate().equals(other.catalog.getP2UrlTemplate())) {
			// syncing only matters when they have the same URL
			return;
		}
		if (Objects.equals(urlOverride, other.urlOverride)) {
			// if they have the same override, that's fine
			return;
		}
		if (urlOverride == null) {
			urlOverride = other.urlOverride;
		} else if (other.urlOverride == null) {
			other.urlOverride = urlOverride;
		} else {
			// they are both non-null and unequal
			throw new IllegalArgumentException(
					catalog.getName()
							+ " "
							+ other.catalog.getName()
							+ " must have the exact same URL, but\n"
							+ urlReasoning()
							+ "\n"
							+ other.urlReasoning());
		}
	}

	private String urlReasoning() {
		if (addedAsTransitiveOf != null) {
			return catalog.getName()
					+ " is using "
					+ url()
					+ " (was added automatically as a transitive of "
					+ addedAsTransitiveOf.catalog.getName()
					+ ")";
		} else {
			return catalog.getName() + " is using " + url();
		}
	}

	public static class TransitiveAwareList<T extends CatalogDsl> {
		private List<CatalogDsl> insertionOrder = new ArrayList<>();
		private final TreeMap<Catalog, CatalogDsl> catalogEntries = new TreeMap<>();

		public void add(T dsl) {
			insertionOrder.add(dsl);
			var existing = catalogEntries.get(dsl.catalog);
			if (existing != null) {
				if (existing.addedAsTransitiveOf != null) {
					throw new IllegalArgumentException(
							dsl.catalog.getName()
									+ " was already added as a transitive dependency of "
									+ existing.addedAsTransitiveOf.catalog.getName()
									+ ".\n"
									+ "You can fix this by moving the <"
									+ dsl.catalog.getName()
									+ "> block above the <"
									+ existing.addedAsTransitiveOf.catalog.getName()
									+ "> block.");
				} else {
					throw new IllegalArgumentException(
							"You can only add " + dsl.catalog.getName() + " once.");
				}
			}
			catalogEntries.put(dsl.catalog, dsl);
			for (var required : dsl.catalog.getRequires()) {
				var transitive = addAsTransitiveOf(required, dsl);
				transitive.syncUrlWith(dsl);
			}
		}

		private CatalogDsl addAsTransitiveOf(Catalog transitive, CatalogDsl originalRequest) {
			var dsl = catalogEntries.get(transitive);
			if (dsl == null) {
				dsl = new CatalogDsl(transitive, originalRequest);
				catalogEntries.put(transitive, dsl);
				insertionOrder.add(dsl);
			}
			for (var required : transitive.getRequires()) {
				var transitiveDsl = addAsTransitiveOf(required, originalRequest);
				transitiveDsl.syncUrlWith(dsl);
			}
			return dsl;
		}

		public void putInto(P2Model model, IdeHook.List hooks) {
			// setup the IDE hooks
			for (CatalogDsl dsl : catalogEntries.values()) {
				model.addP2Repo(dsl.url());
				model.getInstall().addAll(dsl.installs());
				dsl.filters().forEach(model::addFilterAndValidate);
				hooks.addAll(dsl.ideHooks());
			}
			// find the initial perspective
			String perspective =
					insertionOrder.stream()
							.map(dsl -> Catalog.defaultPerspectiveFor(dsl.catalog))
							.filter(Objects::nonNull)
							.findFirst()
							.orElse(null);
			if (perspective != null) {
				IdeHookWelcome welcomeHook =
						(IdeHookWelcome)
								hooks.stream()
										.filter(hook -> hook instanceof IdeHookWelcome)
										.findFirst()
										.orElse(null);
				if (welcomeHook == null) {
					welcomeHook = new IdeHookWelcome();
					hooks.add(welcomeHook);
				}
				if (welcomeHook.perspective() == null) {
					welcomeHook.perspective(perspective);
				}
			}
		}
	}
}
