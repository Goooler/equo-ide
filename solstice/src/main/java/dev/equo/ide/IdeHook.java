/*******************************************************************************
 * Copyright (c) 2023 EquoTech, Inc. and others.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     EquoTech, Inc. - initial API and implementation
 *******************************************************************************/
package dev.equo.ide;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.Map;
import java.util.function.BiConsumer;
import javax.annotation.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * If you add a {@code Bundle-IdeHook} header to the manifest of a jar that points to the fully
 * qualified name of a class that implements this interface, then that class will be instantiated
 * and called at various points in the lifecycle of the IDE.
 */
public interface IdeHook extends Serializable {
	@SuppressWarnings("serial")
	class List extends ArrayList<IdeHook> {
		public List copy() {
			List copy = new List();
			copy.addAll(this);
			return copy;
		}

		InstantiatedList instantiate() {
			var list = new InstantiatedList();
			for (int i = 0; i < size(); ++i) {
				var hookGenerator = get(i);
				try {
					list.list.add(hookGenerator.instantiate());
				} catch (Exception e) {
					list.logError(hookGenerator, e);
				}
			}
			return list;
		}
	}

	class InstantiatedList {
		interface ThrowingConsumer {
			void accept(IdeHookInstantiated hook) throws Exception;
		}

		interface ThrowingBiConsumer<T> {
			void accept(IdeHookInstantiated hook, T arg) throws Exception;
		}

		private final ArrayList<IdeHookInstantiated> list = new ArrayList<>();
		private final Logger logger = LoggerFactory.getLogger(IdeHook.class);
		private @Nullable ArrayList<Map.Entry<Object, Exception>> errorsToReport = new ArrayList<>();
		private @Nullable BiConsumer<Object, Exception> errorLogger;

		private InstantiatedList() {}

		@SuppressWarnings("unchecked")
		@Nullable
		<T extends IdeHookInstantiated> T find(Class<T> clazz) {
			for (var e : list) {
				if (clazz.isInstance(e)) {
					return (T) e;
				}
			}
			return null;
		}

		void logError(Object hook, Exception e) {
			logger.error("IdeHook failure " + hook, e);
			if (errorLogger == null) {
				errorsToReport.add(Map.entry(hook, e));
			} else {
				errorLogger.accept(hook, e);
			}
		}

		/**
		 * Whenever we call IdeHook methods using the `forEach` methods, we catch the exceptions. If an
		 * exception gets thrown, we always log it to the slf4j logger. If an error logger has been set,
		 * then we pass the exception there too. If an error logger has *not* been set, then we store
		 * the exception. When a logger is eventually set, we pass all the exceptions to it.
		 */
		void setErrorLogger(BiConsumer<Object, Exception> errorLogger) {
			if (this.errorLogger != null) {
				try {
					throw new IllegalArgumentException("setErrorLogger was called twice!");
				} catch (IllegalArgumentException e) {
					logger.warn(e.getMessage(), e);
				}
				return;
			}
			this.errorLogger = errorLogger;
			for (var entry : errorsToReport) {
				errorLogger.accept(entry.getKey(), entry.getValue());
			}
			errorsToReport = null;
		}

		void forEach(ThrowingConsumer method) {
			for (IdeHookInstantiated hook : list) {
				try {
					method.accept(hook);
				} catch (Exception e) {
					logError(hook, e);
				}
			}
		}

		<T> void forEach(ThrowingBiConsumer<T> method, T arg) {
			for (IdeHookInstantiated hook : list) {
				try {
					method.accept(hook, arg);
				} catch (Exception e) {
					logError(hook, e);
				}
			}
		}

		Iterator<IdeHookInstantiated> iterator() {
			return list.iterator();
		}
	}

	IdeHookInstantiated instantiate() throws Exception;

	static IdeHookInstantiated usingReflection(String implementationClassName, IdeHook constructorArg)
			throws Exception {
		var clazz = Class.forName(implementationClassName);
		var constructor = clazz.getDeclaredConstructor(constructorArg.getClass());
		return (IdeHookInstantiated) constructor.newInstance(constructorArg);
	}
}
