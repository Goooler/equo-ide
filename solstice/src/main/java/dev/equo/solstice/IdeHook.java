/*******************************************************************************
 * Copyright (c) 2022 EquoTech, Inc. and others.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     EquoTech, Inc. - initial API and implementation
 *******************************************************************************/
package dev.equo.solstice;

public interface IdeHook {
	default void beforeOsgi() {}

	default void afterOsgi(org.osgi.framework.BundleContext context) {}

	default void beforeDisplay() {}

	default void afterDisplay(org.eclipse.swt.widgets.Display display) {}

	/**
	 * This method is called during workbench initialization prior to any windows being opened.
	 *
	 * <p>{@link org.eclipse.ui.application.WorkbenchAdvisor#initialize}
	 */
	default void initialize() {}

	/**
	 * Performs arbitrary actions just before the first workbench window is opened (or restored).
	 *
	 * <p>This method is called after the workbench has been initialized and just before the first
	 * window is about to be opened.
	 *
	 * <p>{@link org.eclipse.ui.application.WorkbenchAdvisor#preStartup}
	 */
	default void preStartup() {}

	/**
	 * Performs arbitrary actions after the workbench windows have been opened (or restored), but
	 * before the main event loop is run.
	 *
	 * <p>{@link org.eclipse.ui.application.WorkbenchAdvisor#postStartup}
	 */
	default void postStartup() {}

	/**
	 * Performs arbitrary finalization before the workbench is about to shut down.
	 *
	 * <p>{@link org.eclipse.ui.application.WorkbenchAdvisor#preShutdown}
	 *
	 * @return <code>true</code> to allow the workbench to proceed with shutdown, <code>false</code>
	 *     to veto a non-forced shutdown
	 */
	default boolean preShutdown() {
		return true;
	}

	/**
	 * Performs arbitrary finalization after the workbench stops running.
	 *
	 * <p>{@link org.eclipse.ui.application.WorkbenchAdvisor#postShutdown}
	 */
	default void postShutdown() {}
}
