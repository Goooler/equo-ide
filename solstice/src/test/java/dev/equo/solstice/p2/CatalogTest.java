/*******************************************************************************
 * Copyright (c) 2023 EquoTech, Inc. and others.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     EquoTech, Inc. - initial API and implementation
 *******************************************************************************/
package dev.equo.solstice.p2;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.diffplug.common.swt.os.SwtPlatform;
import dev.equo.ide.Catalog;
import dev.equo.ide.CatalogDsl;
import dev.equo.ide.IdeHook;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

@ExtendWith({SnapshotExtension.class})
public class CatalogTest {
	private static final TestCatalogDsl JAVA_11_PLATFORM =
			new TestCatalogDsl(Catalog.PLATFORM, "4.27");

	@Test
	public void equoChatGptDefault(Expect expect) {
		expect.toMatchSnapshot(
				requestModel(JAVA_11_PLATFORM, new TestCatalogDsl(Catalog.CHATGPT, null)));
	}

	@Test
	public void equoChatGptFixed(Expect expect) {
		expect.toMatchSnapshot(
				requestModel(JAVA_11_PLATFORM, new TestCatalogDsl(Catalog.CHATGPT, "9.9.9")));
	}

	static class TestCatalogDsl extends CatalogDsl {
		public TestCatalogDsl(Catalog catalog, String urlOverride) {
			super(catalog);
			super.setUrlOverride(urlOverride);
		}
	}

	private String requestModel(TestCatalogDsl... catalogs) {
		var list = new CatalogDsl.TransitiveAwareList<TestCatalogDsl>();
		for (var catalog : catalogs) {
			list.add(catalog);
		}
		var model = new P2Model();
		var hooks = new IdeHook.List();
		list.putInto(model, hooks);

		String result = ConsoleTable.request(model, ConsoleTable.Format.ascii);
		for (SwtPlatform platform : SwtPlatform.getAll()) {
			result = result.replace(platform.toString(), "platform-specific");
		}
		return result;
	}
}
