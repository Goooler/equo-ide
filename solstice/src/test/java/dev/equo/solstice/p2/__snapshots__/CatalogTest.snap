dev.equo.solstice.p2.CatalogTest.equoChatGptDefault=[
kind,value
p2repo,https://download.eclipse.org/eclipse/updates/4.27/
install,org.eclipse.platform.ide.categoryIU
maven,com.equo:com.equo.chromium.cef.platform-specific:106.0.10
maven,com.equo:com.equo.chromium:106.0.10
maven,dev.equo.ide:equo-ide-chatgpt:1.0.1

]


dev.equo.solstice.p2.CatalogTest.equoChatGptFixed=[
kind,value
p2repo,https://download.eclipse.org/eclipse/updates/4.27/
install,org.eclipse.platform.ide.categoryIU
maven,com.equo:com.equo.chromium.cef.platform-specific:106.0.10
maven,com.equo:com.equo.chromium:106.0.10
maven,dev.equo.ide:equo-ide-chatgpt:9.9.9

]